src/storage-client/src/storage_collections.rs

```diff
@@ -1453,11 +1453,7 @@ where
         // Reorder in dependency order.
         to_register.sort_by_key(|(id, ..)| *id);
 
-        // We hold this lock for a very short amount of time, just doing some
-        // hashmap inserts and unbounded channel sends.
-        let mut self_collections = self.collections.lock().expect("lock poisoned");
```

Comment:
To me, it felt safer that we hold the lock for the whole loop, which avoids inconsistencies in collections/dependencies sneaking in. That might be me being overly cautious, though.

Reply:
The reason I broke this up is because it's not possible to hold this lock while also doing async work. Definitely open to other ways of structuring this as long as they compile!

Thinking out loud a bit:
- The lock is held twice for each `GlobalId`: once briefly to find the dependencies, and again to actually perform the mutations.
    - The first lock only actually performs reads; I've pushed a small commit that removes some unneeded `mut`s to prove this to the compiler.
- So for each GID, all the changes to `self_collections` take place atomically. But in theory some concurrent code could observe some collections having been created but not others. (And this includes eg. only some subsources of a particular source being available.)
    - I don't have any particular reason to fear this, but your instincts are certainly more finely honed than mine here... any specifics that give you pause?

Reply:
My fear is that, once we've released the lock, our changes are available to other callers who may be accessing the `StorageCollections` object concurrently. So there's a (probably very narrow) window where the shard is visible to others but has undesirable frontiers.

I'm not sure how much to worry about this in practice... maybe there are higher-level things that ensure this weird state can never be observed? But it seems to me less safe than getting the shard into the correct state before it's visible to callers of `StorageCollections`, even if it means all the full set of collections isn't registered atomically.